File: web/src/utils/scoring.ts
export function roundScore(value: number): number {
  return Math.round(value * 100) / 100
}

export function calculateScores(scores: (number | null)[]) {
  if (scores.some((s) => s === null)) {
    return { avgValues: null, avgCapability: null, avgOutput: null, finalScore: null }
  }
  const s = scores as number[]
  const avgValues = roundScore((s[0] + s[1] + s[2]) / 3)
  const avgCapability = roundScore(s.slice(3, 10).reduce((a, b) => a + b, 0) / 7)
  const avgOutput = roundScore((s[10] + s[11]) / 2)
  const finalScore = roundScore(avgValues * 0.2 + avgCapability * 0.4 + avgOutput * 0.4)
  return { avgValues, avgCapability, avgOutput, finalScore }
}

export function suggestResult(finalScore: number): { sys: string; result: string | null } {
  if (finalScore <= 2) return { sys: '不通过', result: '不通过晋升' }
  if (finalScore >= 4) return { sys: '通过', result: '通过晋升' }
  return { sys: '评委自选', result: null }
}

File: backend/app/scoring.py
from decimal import Decimal, ROUND_HALF_UP


def round_score(value: float) -> float:
    return float(Decimal(str(value)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP))


def calculate_scores(scores: list[int | None]) -> dict[str, float | None]:
    if any(s is None for s in scores):
        return {
            "avg_values": None,
            "avg_capability": None,
            "avg_output": None,
            "final_score": None,
        }

    avg_values = round_score(sum(scores[0:3]) / 3)
    avg_capability = round_score(sum(scores[3:10]) / 7)
    avg_output = round_score(sum(scores[10:12]) / 2)
    final_score = round_score(avg_values * 0.2 + avg_capability * 0.4 + avg_output * 0.4)

    return {
        "avg_values": avg_values,
        "avg_capability": avg_capability,
        "avg_output": avg_output,
        "final_score": final_score,
    }
